### PR TITLE
Optimize rebuffering

### PR DESCRIPTION
**Pull request by kiryk**
This PR optimizes `RepairSetup::rebufferBottomUp` with three types of changes:
1. Vectors are preallocated wherever their size can be predicted
2. Slack values, used in sorting, are precalculated.
3. Option pruning is modified from pessimistically quadratic to linear complexity.

I gathered averaged outcomes from 10 runs on chosen designs (`cts` target):
 Version \ Design | Ariane | BlackParrot | TinyRocket | Ibex
-|-|-|-|-
Unoptimized | 63m40s | 3m39s | 0m27s | 1m24s
Optimized | 12m32s | 2m57s | 0m21s | 0m28s

### CODE REVIEW

**Comment by maliberty** [comment]
I see these failures
```
array_ins_delay (tcl) *ERROR* segmentation violation
insertion_delay (tcl) *ERROR* segmentation violation
```
they were broken in master a while ago and then fixed.  I suspect you need to merge master into your branch.

**Comment by QuantamHD** [comment]
Wow! Amazing work

File: src/rsz/src/Rebuffer.cc [hunk 1]
```diff
@@ -305,8 +306,10 @@ RepairSetup::addWireAndBuffer(const BufferedNetSeq& Z,
                               int level)
 {
   BufferedNetSeq Z1;
+  Z1.resize(Z.size());
   Point wire_end = bnet_wire->location();
-  for (const BufferedNetPtr& p : Z) {
+  for (int i = 0; i < Z.size(); i++) {
+    const BufferedNetPtr& p = Z[i];
```
[the remark this answers is not in the document]
Comment (kiryk) [reply src/rsz/src/Rebuffer.cc]: Nice, right.

File: src/rsz/src/Rebuffer.cc [hunk 2]
```diff
@@ -227,55 +240,44 @@ RepairSetup::rebufferBottomUp(const BufferedNetPtr& bnet,
...
+        slacks[junc.get()] = slackPenalized(junc);
+        Z[base + j] = std::move(junc);
       }
     }
     // Prune the options if there exists another option with
     // larger required and smaller capacitance.
-    // This is fanout^2.
-    // Presort options to hit better options sooner.
+    // This is fanout*log(fanout) if options are
+    // presorted to hit better options sooner.
```
Comment (maliberty) [inline src/rsz/src/Rebuffer.cc:250]: I think both comments are wrong.  If the fanouts are evenly split in two between Z1 & Z2 then Z is size (fanout/2)^2.  The sort is big-O N*logN of N=fanout^2 in that case.

File: src/rsz/src/Rebuffer.cc [hunk 3]
```diff
@@ -227,55 +240,44 @@ RepairSetup::rebufferBottomUp(const BufferedNetPtr& bnet,
...
-        const BufferedNetPtr& q = Z[qi];
-        float Lq = q->cap();
-        // We know Tq <= Tp from the sort so we don't need to check req.
-        // If q is the same or worse than p, remove solution q.
-        if (fuzzyLess(Lq, Lp)) {
-          // Copy survivor down.
-          Z[si++] = q;
-        }
+      // If Tp is the same or worse than Tsmall, remove solution p.
+      if (fuzzyLess(Lp, Lsmall)) {
```
Comment (maliberty) [inline src/rsz/src/Rebuffer.cc:274]: The comment doesn't match the code Tp vs Lp and Tsmall vs Lsmall.

File: src/rsz/src/Rebuffer.cc [hunk 4]
```diff
@@ -227,55 +240,44 @@ RepairSetup::rebufferBottomUp(const BufferedNetPtr& bnet,
...
+        // Otherwise copy the survivor down.
+        Z[si++] = p;
+        Lsmall = Lp;
```
Comment (maliberty) [inline src/rsz/src/Rebuffer.cc:277]: The previous code did not update Lp so it likely kept many more options.  This is a change in functionality not a pure performance tuning and will require validation if is intended - please confirm.   It sounds reasonable.  I expect this is where most of the performance gain comes from.

File: src/sta
Comment (maliberty) [inline src/sta:1]: Is this related to this change?

**Review by maliberty — CHANGES_REQUESTED** [review]

**Comment by maliberty** [comment]
Started a secure CI

File: src/rsz/src/Rebuffer.cc (hunk 2 above)
Comment (kiryk) [reply src/rsz/src/Rebuffer.cc:250]: The complexity is a function of two variables here, but I decided to keep the original intent. Hm, I think we can just avoid the issue saying that the problem is N*logN, without any explicit reference to fanout.

File: src/rsz/src/Rebuffer.cc (hunk 4 above)
Comment (kiryk) [reply src/rsz/src/Rebuffer.cc:277]: Yes, it is intended. I changed it because an improvement of complexity seemed reasonable, but in the designs I tested most of improvement came not from this, but from caching the slack values (and so avoiding redundant floating point calculations).

File: src/sta
Comment (maliberty) [reply src/sta:1]: I see a bunch of failures that look related to this change in the CI.  I think you should back it out of this PR.

Comment (maliberty) [reply src/sta:1]: I have reverted it in my copy of this and restarted the CI

File: src/rsz/src/Rebuffer.cc [hunk 5]
```diff
@@ -216,9 +216,22 @@ RepairSetup::rebufferBottomUp(const BufferedNetPtr& bnet,
...
-    for (const BufferedNetPtr& p : Z1) {
-      for (const BufferedNetPtr& q : Z2) {
+
+    size_t size = Z1.size()*Z2.size();
+    // This assumption is used only in the final loop
+    // but we can quit as early as we know there's nothing to do here.
+    if (size == 0) {
+      return Z;
+    }
+    Z.resize(size);
```
[the remark this answers is not in the document]
Comment (kiryk) [reply src/rsz/src/Rebuffer.cc]: Done.

File: src/rsz/src/Rebuffer.cc [hunk 6]
```diff
@@ -227,55 +240,44 @@ RepairSetup::rebufferBottomUp(const BufferedNetPtr& bnet,
...
+    // This is fanout*log(fanout) if options are
+    // presorted to hit better options sooner.
     sort(Z.begin(),
          Z.end(),
-         [this](const BufferedNetPtr& option1, const BufferedNetPtr& option2) {
-           Slack slack1 = slackPenalized(option1);
-           Slack slack2 = slackPenalized(option2);
+         [&slacks](const BufferedNetPtr& option1, const BufferedNetPtr& option2) {
+           Slack slack1 = slacks[option1.get()];
+           Slack slack2 = slacks[option2.get()];
```
[the remark this answers is not in the document]
Comment (kiryk) [reply src/rsz/src/Rebuffer.cc]: Done.

**Comment by maliberty** [comment]
There was some issue in the secure CI with the SQL server going down while this was running but as far as I can see everything passed.

File: src/rsz/src/Rebuffer.cc (hunk 4 above)
Comment (kiryk) [reply src/rsz/src/Rebuffer.cc:275]: The previous code would iterate over the vector and for each element it'd remove its successors with a higher or similar capacitance. If some element is not removed, it must have a lower capacitance, so on the next iteration the inner loop will iterate over the survivors again and challenge them even more. It should be enough to iterate once, while keeping track of the lowest cap so far.

I think If this argument is correct, then both versions should produce the same output.

**Comment by QuantamHD** [comment]
This ready to go @maliberty?
